pkg/authentication/scramcredentials/scram_credentials.go

```diff
@@ -0,0 +1,192 @@
+package scramcredentials
+
+import (
+	"crypto/hmac"
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"hash"
+
+	"github.com/xdg/stringprep"
+)
+
+const (
+	RFC5802MandatedSaltSize = 4
+
+	clientKeyInput = "Client Key" // specified in RFC 5802
+	serverKeyInput = "Server Key" // specified in RFC 5802
+
+	// using the default MongoDB values for the number of iterations depending on mechanism
+	scramSha1Iterations   = 10000
+	scramSha256Iterations = 15000
+
+	sha256mechanismName = "SCRAM-SHA-256"
+	// MONGODB-CR is an umbrella term for SCRAM-SHA-1 and MONGODB-CR for legacy reasons, once MONGODB-CR
+	// is enabled, users can auth with SCRAM-SHA-1 credentials
+	sha1MechanismName = "MONGODB-CR"
+)
+
+type ScramCreds struct {
+	IterationCount int    `json:"iterationCount"`
+	Salt           string `json:"salt"`
+	ServerKey      string `json:"serverKey"`
+	StoredKey      string `json:"storedKey"`
+}
+
+func ComputeScramSha256Creds(username, password string, salt []byte) (ScramCreds, error) {
+	return computeCreds(username, password, salt, sha256mechanismName)
```

Comment:
this function only dispatches the request to `computeCreds` and then, there's branching in `computeCreds` depending on that..

Why not:

```golang
func ComputeScramSha256Creds(username, password string, salt []byte) (ScramCreds, error) {
	base64EncodedSalt := base64.StdEncoding.EncodeToString(salt)
	return computeScramCredentials(sha256.New, scramSha256Iterations, base64EncodedSalt, password)
}

func ComputeScramSha1Creds(username, password string, salt []byte) ScramCreds, error) { 
	base64EncodedSalt := base64.StdEncoding.EncodeToString(salt)
	password = md5Hex(username + ":mongo:" + password)
	return computeScramCredentials(sha1.New, scramSha1Iterations, base64EncodedSalt, password)
}
```

Reply:
this is actually much nicer, as the Sha256 function doesn't actually require `username` 👍